Make the priority for picking the storage driver configurable

dcermak opened the pull request:
This fixes https://github.com/containers/storage/issues/1457

I have tested this locally by vendoring the specific PR into podman and then setting the following `~/.config/containers/storage.conf`:
```toml
[storage]

driver_priority = [
                "zfs",
                "btrfs",
                "overlay"
]
```
podman then tried zfs (but failed as I don't have any zfs partitions) and picked btrfs as the next viable candidate.

Removing the setting causes podman to correctly pick up overlay as the driver.

rhatdan commented:
Please update the man page and make it clear the priority list is only used if graphdriver name is NOT specified.

rhatdan commented on docs/containers-storage.conf.5.md:
```diff
@@ -57,7 +57,12 @@ A common use case for this field is to provide a local storage directory when us
 **runroot**=""
   container storage run dir (default: "/run/containers/storage")
-Default directory to store all temporary writable content created by container storage programs. The rootless runroot path supports environment variable substitutions (ie. `$HOME/containers/storage`)
+Default directory to store all temporary writable content created by container storage programs. The rootless runroot path supports environment variable substitutions (ie. `$HOME/containers/storage**)
```
This change looks like a mistake.

dcermak replied:
Indeed, thanks for the catch!

rhatdan commented:
One issue in the man page, once this is fixed, LGTM
@vrothberg PTAL

dcermak commented:
One question about this change though: when invoking podman now, I **always** get the following error:
```
ERRO[0000] The storage 'driver' option must be set in /home/dan/.config/containers/storage.conf to guarantee proper operation
```
which is also suboptimal from a UX perspective. Does it really need to be an error? The driver is after all set in the store itself.

vrothberg commented:
> One question about this change though: when invoking podman now, I **always** get the following error:
> 
> ```
> ERRO[0000] The storage 'driver' option must be set in /home/dan/.config/containers/storage.conf to guarantee proper operation
> ```
> 
> which is also suboptimal from a UX perspective. Does it really need to be an error? The driver is after all set in the store itself.

@giuseppe @rhatdan WDYT?  I am surprised as well.  Shouldn't Podman / storage be smart enough to pick a default driver if none is set?

vrothberg commented on drivers/driver.go:
```diff
@@ -329,15 +329,22 @@ type Options struct {
 
 // New creates the driver and initializes it at the specified root.
-func New(name string, config Options) (Driver, error) {
+func New(name string, driverPriority []string, config Options) (Driver, error) {
```
This is a breaking chance.  Please move it to the `Options` struct.

vrothberg commented on types/options.go:
```diff
@@ -213,10 +214,16 @@ type StoreOptions struct {
-	// using.  It only needs to be specified the first time a Store is
-	// initialized for a given RunRoot and GraphRoot.
+	// If the driver is not specified, the best suited driver will be picked
+	// either from GraphDriverPriority or from the platform dependent
```
```suggestion
	// either from GraphDriverPriority, if specified, or from the platform dependent
```

vrothberg commented on docs/containers-storage.conf.5.md:
```diff
@@ -59,6 +59,11 @@ A common use case for this field is to provide a local storage directory when us
 Default directory to store all temporary writable content created by container storage programs. The rootless runroot path supports environment variable substitutions (ie. `$HOME/containers/storage`)
 
+**driver_priority**=[]
+  Priority list for the storage drivers that will be tried in case no driver is set.
```
"will be tried" is a bit vague.  Please be specific when a certain driver will be used and when not.

dcermak replied:
Does it make more sense now?

dcermak commented:
> > One question about this change though: when invoking podman now, I **always** get the following error:
> > ```
> > ERRO[0000] The storage 'driver' option must be set in /home/dan/.config/containers/storage.conf to guarantee proper operation
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > which is also suboptimal from a UX perspective. Does it really need to be an error? The driver is after all set in the store itself.
> 
> @giuseppe @rhatdan WDYT? I am surprised as well. Shouldn't Podman / storage be smart enough to pick a default driver if none is set?

The error comes from here fyi: https://github.com/containers/storage/blob/46e94559ce9dec8bf0ba4fb04c2df4664bb728f5/types/options.go#L384 and I suspect it has been there for a while (git blame unfortunately only goes back to a commit from 2021)

giuseppe commented on drivers/driver.go:
```diff
@@ -337,7 +338,14 @@ func New(name string, config Options) (Driver, error) {
+	}
+
+	for _, name := range prioList {
 		if name == "vfs" {
```
if vfs is set in the `config.DriverPriority` then I expect it to be honored.

I think it should be changed to `name == "vfs" && len(config.DriverPriority) == 0`

giuseppe reviewed (commented):
could we keep the error only when DriverPriority is empty (and maybe change it to a warning)?

rhatdan commented:
I believe at some point we forced users to specify their driver, because they kept on failing over to VFS in rootless mode, and then complained when they ran out of disk space in their homedir.

Forcing them to specify overlay and failing, caused us to know to have them install fuse-overlayfs before the kernel supported rootless overlay.

Now that rootless overlay will work for almost everyone the overlay default should work, and we might not need to require them to enter any longer.

rhatdan commented:
@dcermak Would like to get this in as we are about to release a new version of containers/storage.  Are you able to address the comments?

dcermak commented:
> could we keep the error only when DriverPriority is empty (and maybe change it to a warning)?

Given @rhatdan's comment for the historical context, I'd honestly suggest to remove this error completely. WDYT?

rhatdan commented:
Could we do the check iff driver is not specified and priority is not specified, then fail.

dcermak commented:
> Could we do the check iff driver is not specified and priority is not specified, then fail.

We do have an internal priority list as a fallback and also in `New` we try more or less every driver until one of them works: https://github.com/containers/storage/blob/fc91849352e5de23e94a01a659131355e3a31017/drivers/driver.go#L332

Thus I don't consider that warning really valid or is there a case where the user really should define a driver?

rhatdan commented:
I am not crazy about the internal list, since it is not easily understood by humans, so I would rather fail and tell them (or the distribution, specify one rather then having it randomly chosen potentially based on which software they have installe.

If the user or distribution specifies a search list of the driver then we should not warn.

dcermak commented:
@rhatdan Makes sense, I've adjusted the code accordingly, modified the error to be just a warning and tweaked the message as well to sound less scary.

rhatdan commented:
LGTM
@vrothberg @giuseppe PTAL

giuseppe reviewed (approved):
LGTM
